### DISCUSSION TITLE
Specify different cache configuration in runtime

### DISCUSSION THREAD

**Discussion by mabartos** [keycloak/keycloak#24055] [Ideas]
We've encountered requirements to specify the cache configuration file for various different deployments for an already built image. At this moment, the specifying of the ISPN configuration can be done only in build time, and it's not possible to change it. However, our build steps execution for the configuration handling is made in runtime with the help of recorded bytecode, but still; in runtime.

It means, we've marked these cache options as build time, but we don't leverage the benefits of processing them during the build time. 
The overall parsing of the configuration is quite expensive. I've tried to profile it, and only the parsing takes ~34 CPU samples during the startup, where the execution of it may take more than ~230ms and a certain amount of additional system resources for a simple config file.

In a prod environment, the time spent on the parsing might be negligible to the execution of process steps for other components of the system. The possibility of changing the configuration file in runtime might be probably preferred in this context.  

The initialization of the CacheManager is happening during runtime anyway, so I don't see any problems providing the possibility to override the cache config file even in runtime execution.

We could make improvements to conform to both requirements. There might be several approaches to tackle that:

**Current approach**
1. Option `cache-config-file` can be specified **only in BUILD time**
2. Expensive parsing done in **RUNTIME** (bad)
3. **Impossible** to change the config file without reaugmentation (bad)

**a) The simplest improvement**
1. Option `cache-config-file` can be specified in **BUILD/RUNTIME**
2. Expensive parsing done in **RUNTIME** (bad)
3. **Possible** to change the config file without reaugmentation

**b) More efficient way with simple changes**
1. Option `cache-config-file` can be specified in **BUILD/RUNTIME**
2. Expensive parsing done in **BUILD time**
3. Store a hash of the read configuration file to the build config 
4. If the prop is specified in **RUNTIME**:
     4.1) has a different name: parse the configuration in **RUNTIME**
     4.2) has the same name, and the hash is different: parse the configuration in **RUNTIME**
5. **Possible** to change the config file without reaugmentation

**c) Most efficient way (in certain context) with more changes**
1. Option `cache-config-file` can be specified in **BUILD/RUNTIME**
2. User can specify more config files separated by a comma (`--cache-config-file config1.xml,config2.xml`) (or option `cache-config-files`)
3. Expensive parsing done in **BUILD** time
4. During **RUNTIME**, specify only one config file, which is already parsed and used (`--cache-config-file config2.xml`)
5. Get rid of the additional parsed ISPN configuration objects (more memory allocations :/)
6. No expensive parsing done in **RUNTIME**
7. **Possible** to change the config file without reaugmentation
---------------
I'm leaning towards the **b)** option as it provides the required functionality in quite an efficient way. There will not be any breaking changes, only an improvement to the current functionality. 

@vmuzikar @Pepo48 @shawkins @pedroigor WDYT?
cc: @rh-iatanaso

**Comment by shawkins** [comment]
> a) The simplest improvement
> b) More efficient way with simple changes

This doesn't seem easily doable.  Currenlty options can be runtime or buildtime, but not both.  I think that reinforces the quarkus best practice.  So for option a) that would probably mean the simpliest approach is just changing the option to a runtime one - which would be a breaking change right?

**Reply by mabartos** [reply]
As stated in Quarkus docs[1] for the build options:

>However, some extensions do define properties overridable at runtime. A simple example is the database URL, username and password which is only known specifically in your target environment, so they can be set and influence the application behaviour at runtime.

In the end, these properties overridable at runtime are considered as the runtime properties. We could consider it as well, but basically accept the option in build time and also in runtime. Based on it, we could react to the value in the build time. 

@shawkins Or do you think it could break some best-practices/conventions around it?

EDIT: It's probably not ideal and we could have a unified solutions for these options.  

[1] https://quarkus.io/guides/config-reference#build-time-configuration

**Reply by mabartos** [reply]
> So for option a) that would probably mean the simpliest approach is just changing the option to a runtime one - which would be a breaking change right?

Yes, that would be a breaking change due to the validation of these options.

**Reply by shawkins** [reply]
> Or do you think it could break some best-practices/conventions around it?

It would be best to not to.  Things got confusing quickly with a similar issue in the josdk, and allowing options / propertymappers to be both runtime and buildtime may require a lot of changes.

> c) Most efficient way (in certain context) with more changes

Based upon the above this would need to split into two properties --cache-config-file (runtime, and a breaking change) and --cache-config-files (buildtime).  Would it be possible to discovery cache config files in the /conf so that --cache-config-files is not needed?

**Reply by mabartos** [reply]
> It would be best to not to. Things got confusing quickly with a similar issue in the josdk, and allowing options / propertymappers to be both runtime and buildtime may require a lot of changes.

@shawkins Yep, agree, rather not doing it.

> Would it be possible to discovery cache config files in the /conf so that --cache-config-files is not needed?

Good point; I like it. The `/conf` dir might contain more config files, but we don't support XML configuration for something else than the ISPN config. Thus we can parse all XML files (or we'd need to specify some prefix for these files) in build time and if it's not ISPN config, the build fails. The specific ISPN configuration will be chosen during the runtime via the `--cache-config-file` option. 

---------------------
So, the proposed approach might look like this:

1. Gather ISPN configuration files (XML files) included in the `/conf` dir 
2. Parse them in **BUILD** time
3. Get the value of the `--cache-config-file` option in **RUNTIME**
     a) If we already have the file parsed, we use it
     b) If it's a new ISPN configuration, parse it in **RUNTIME**
4. **Possible** to change the config file without reaugmentation

@shawkins @vmuzikar @Pepo48 WDYT?

**Reply by shawkins** [reply]
Sounds good to me.  That would just be a small breaking change.

**Reply by vmuzikar** [reply]
@mabartos Thanks for the analysis.

I have to agree with @shawkins that we should not mix build and runtime options as it might lead to some confusing behaviour.

I kinda like the idea of leveraging a config dir for storing build time config file. It makes most sense as you package it together with your optimized Keycloak.

I would not introduce additional `--cache-config-files` option as it might be a bit ambiguous how it works with `--cache-config-file`.

I'd suggest the following iterative approach:
1. Make `--cache-config-files` run time option.
2. Introduce a well known location for a single ISPN config file (`{kc_home}/conf/cache-ispn.xml` which we actually already have). If this file is present, it's processed build time. `--cache-config-files` takes precedence if specified and is still processed run time.

WDYT?

I think we should also involve @keycloak/store in the discussion.

**Reply by mabartos** [reply]
@vmuzikar I assume you meant the `--cache-config-file` (without the 's') in the suggested approach. 

> Introduce a well known location for a single ISPN config file. If this file is present, it's processed build time. `--cache-config-file` takes precedence if specified and is still processed run time.

Yes, the default value for the `--cache-config-file` is the `cache-ispn.xml` (when `--cache=ispn`), so nothing would change. Yes, the option takes precedence over the default ISPN config file, and the new config is used; as the current approach. 

I just want to confirm we're on the same page. Have I explained it sufficiently that all ISPN configuration files included in the `conf/` dir would be parsed in build time and stored and after that, you don't have to parse them in runtime? For instance, you have `conf/cache-ispn-prod1.xml` and `conf/cache-ispn-prod2.xml` ISPN config files. Both of them are parsed in build time and stored as MultiBuildItem in the build chain. In runtime, when we use the `--cache-config-file=cache-ispn-prod2.xml`, only the second config is used for the ISPN, and the first one is freed up from memory. 

@vmuzikar Or do you have any concerns with that? 

I'd probably provide a specific prefix(`cache-ispn-*.xml`) for the ISPN config files in order to prevent certain issues in some custom configuration files used in customers' deployments.

**Reply by mhajas** [reply]
Thank you for bringing this up @mabartos! I think the idea to move the parsing to build time is correct and will definitely increase startup times.

I am not sure I understand correctly one thing. What would happen if someone specifies different `--cache-config-file=` in runtime? Would it fail? Or we would parse the new configuration? 

I am a little bit concerned about parsing all xml files in `{kc_home}/conf` dir. Even though we have there only ISPN configurations currently, it can change in the future. 

Can we have a new directory called `additional_ispn_conf` or similar in `{kc_home}/conf` directory? Then it would make sense to me to **not** have `--cache-config-files` property at all, parse all xml files in `{kc_home}/conf/additional_ispn_conf` + ``{kc_home}/conf/cache-ispn.xml` and then in runtime pick the correct configuration based on `--cache-config-file`. I would maybe even refuse to parse additional configurations at that point but not sure whether it is not too breaking. 

WDYT?

**Reply by mabartos** [reply]
> I am not sure I understand correctly one thing. What would happen if someone specifies different --cache-config-file= in runtime? Would it fail? Or we would parse the new configuration?

@mhajas If the file wasn't parsed in build time, it'll be parsed in runtime; the same as the current approach in Keycloak.  

> I am a little bit concerned about parsing all xml files in {kc_home}/conf dir. Even though we have there only ISPN configurations currently, it can change in the future.

Yep, it concerns me as well, as it might not be a long-term retained solution. 

> Can we have a new directory called additional_ispn_conf or similar in {kc_home}/conf directory?

I'm not sure if we should split the ISPN configuration files. I'll try to add some possible approaches in the next comment based on your review.

**Reply by mabartos** [reply]
Comparison of possible approaches:

**a) Specific directory**

Considerations:
- The dir should be inside of the `conf/` to have one central point for configuration
- IMO, the default ISPN configuration and the additional ones should be clustered together in one dir
- If they're together in a separate directory, it'd require changes to the location to the default config file from the codebase (probably not a big deal)

Proposal:

- create a `conf/infinispan/` dir with the default ISPN config file, where users could add their own custom ISPN config files, which would be parsed in build time
- the subdir might contain brief README with clear intention and usage of it
- in runtime, users can specify directly the name of the file, which is present in the `conf/infinispan/` dir
- in runtime, users can specify an absolute path to the file; check if parsed in build time, otherwise parse it in runtime

By providing the `infinispan/` directory, we'd separate the configuration for the ISPN-only as there might be various different setups needed (i.e. Cross-DC). The other configurations would be present in the general `conf/` dir (f.e. `keycloak.conf` file).

_Evaluation:_
(+) config separation
(+) we know what files should be parsed
(+) easy understanding of changes, benefits, and usage (with a provided README)
(-) changes to the location of the default ISPN config file in the codebase   

**b) Specific directory with the additional configs (Michal's suggestion)**

- create a `conf/additional_ispn_conf/` dir with additional ISPN config files
- the default `cache-ispn.xml` would stay in the `conf/` dir
- parse the default and the additional ISPN config files in buildtime 

_Evaluation:_
(+) we know what files should be parsed
(+) easy understanding of changes, benefits, and usage (with a provided README to the `additional_ispn_config`?)
(+) NO changes to the location of the default ISPN config file in the codebase   
(-) ISPN config files on different locations (might be confusing)

**c) Specific prefix**

-  We could gather only XML files with a specific prefix that we know it's an ISPN configuration (`cache-ispn-*.xml`)
-  No flexibility; users need to rename their files

_Evaluation:_
(+) we know what files should be parsed
(+) no migration needed
(-) not flexible

**d) Parse only Infinispan XML files**

- Gather all XML files from the `conf/` dir
- Check the used schema/root for the XML file
- If the ISPN schema/root(`<infinispan/>`) is used, parse it in build time
- Otherwise, skip the XML parsing. 

_Evaluation:_
(+) no migration needed
(+) flexible in terms of names
(-) we do not know what files should be parsed yet
(-) get the name of the schema/root-element

---------------------------
I'm kinda leaning towards the **a)** option as the manageability of the additional config files would be quite straightforward and we would have component-separated configuration. The only drawback is that we'd need to change the location of the default config file in the codebase. 

@mhajas @shawkins @vmuzikar WDYT?

**Reply by mhajas** [reply]
The disadvantage of **a)** is that it will result in a migration step for each Keycloak deployment. Upgrading Keycloak includes extracting the new version and then copying old `conf/` and others to the new location. This means the procedure will result in two XML files which may be confusing if people don't read properly the migration guide (maybe they will skip Infinispan steps because they are using the default config and assume everything will work). On the other hand, option **a)** is cleaner. 

I like option **b)** more but I don't insist on it. Option **a)** works for me too.

**Reply by vmuzikar** [reply]
For me option a) feels most cleanest. Yes, migration steps are involved but it's for a future good. :)

> * in runtime, users can specify an absolute path to the file; check if parsed in build time, otherwise parse it in runtime

For simplicity sake, I'd avoid allowing absolute paths (also for reducing possible attack surface). I'd limit it strictly to the ISPN config dir. I.e. `--cache-config-file` accepts a filename only, no paths involved (relative or absolute). Yeah, it's probably another breaking change...

We should  also define behaviour what happens when `--optimized` is used and the config file (either currently used or unused) was changed or even deleted.

**Reply by mabartos** [reply]
> For me option a) feels most cleanest. Yes, migration steps are involved but it's for a future good. :)

We might have some sort of hybrid approach related to the default ISPN config, but I'd like to avoid providing any more advanced logic behind it. I'd prefer to assume users read the migration guide properly :) And if the ISPN config file is not found, we could show an error message such as "_Cannot find the default Infinispan config file. Did you move it to the 'conf/infinispan/' directory?_"

> For simplicity sake, I'd avoid allowing absolute paths (also for reducing possible attack surface). I'd limit it strictly to the ISPN config dir. I.e. --cache-config-file accepts a filename only, no paths involved (relative or absolute). Yeah, it's probably another breaking change...

+1

> We should also define behaviour what happens when --optimized is used and the config file (either currently used or unused) was changed or even deleted.

During the processing of the file in build time, we can store also the `lastModified` attribute of the file. In runtime, we can easily check if the `lastModified` attribute is different and if yes, display an INFO message "_the ISPN config file has been changed and will be parsed again. Consider executing the build phase again with the new changes..._".

As we discussed the case when it's possible to parse an unknown ISPN config file in runtime, I'd consider it a reasonable approach.

@mhajas @vmuzikar WDYT?

**Reply by vmuzikar** [reply]
Hashes would be probably better than a lastModified timestamp. However, we'd need to calculate the hash at every server start incl. when `--optimized` is used. That's not optimal, so timestamp it is then. But instead of INFO log message, I'd fail the server start.

**Comment by rh-iatanaso** [comment]
> For me option a) feels most cleanest. Yes, migration steps are involved but it's for a future good. :)
> 
> > * in runtime, users can specify an absolute path to the file; check if parsed in build time, otherwise parse it in runtime
> 
> For simplicity sake, I'd avoid allowing absolute paths (also for reducing possible attack surface). I'd limit it strictly to the ISPN config dir. I.e. `--cache-config-file` accepts a filename only, no paths involved (relative or absolute). Yeah, it's probably another breaking change...

I am not sure what are the security issues with absolute paths, but relative paths are an easy way to introduce backward compatibility issues if the default path is changed. 
Absolute path does make it easier for user managed resources. `keycloak.conf ` supports absolute path. Also depending on how the content of the file is delivered it might not be even possible to add path in the relative location. (e.g content comes from ConfigMap or something similar)

> We should also define behaviour what happens when `--optimized` is used and the config file (either currently used or unused) was changed or even deleted.
